scenariolib/uabot.go

```diff
@@ -0,0 +1,154 @@
+	local          bool
+	scenarioURL    string
+	searchToken    string
+	analyticsToken string
+	random         *rand.Rand
+}
+
+func NewUabot(local bool, scenarioUrl string, searchToken string, analyticsToken string, random *rand.Rand) *uabot {
+	return &uabot{
+		local,
+		scenarioUrl,
+		searchToken,
+		analyticsToken,
+		random,
+	}
+}
+
+func (bot *uabot) Run() error {
+	var (
+		conf *Config
+		err error
+		timeVisits int
+	)
+
+	// Init from path instead of URL, for testing purposes
+	if bot.local {
+		conf, err = NewConfigFromPath(bot.scenarioURL)
+	} else {
+		conf, err = NewConfigFromURL(bot.scenarioURL)
+	}
+	if err != nil {
+		return err
+	}
+
+	// Refresh the scenario files every 5 hours automatically.
+	// This way, no need to stop the bot to update the possible scenarios.
+	bot.continuallyRefreshScenariosEvery(5 * time.Hour, conf)
+	if conf.TimeBetweenVisits > 0 {
+		timeVisits = conf.TimeBetweenVisits
+	} else {
+		timeVisits = DEFAULTTIMEBETWEENVISITS
+		Info.Println("ajalwkdjalwkjdalkwjdalwkdjwalkdjalwkjd")
+		bot.continuallyUpdateTimeVisitsEvery(24 * time.Hour, &timeVisits)
+	}
+
+	count := 0
+	for {
+		// Run forever
+
+		scenario, err := conf.RandomScenario()
+		if err != nil {
+			return err
+		}
+
+		if scenario.UserAgent == "" {
+			scenario.UserAgent, err = conf.RandomUserAgent(false)
+			if err != nil {
+				return err
+			}
+		}
+
+		// New visit
+		visit, err := NewVisit(bot.searchToken, bot.analyticsToken, scenario.UserAgent, conf)
+		if err != nil {
+			return err
+		}
+
+		// Setup specific stuff for NTO
+		//visit.SetupNTO()
+		// Use this line instead outside of NTO
+		visit.SetupGeneral()
+		visit.LastQuery.CQ = conf.GlobalFilter
+
+		err = visit.ExecuteScenario(*scenario, conf)
+		if err != nil {
+			return err
+		}
+
+		visit.UAClient.DeleteVisit()
+		time.Sleep(time.Duration(bot.random.Intn(timeVisits)) * time.Second)
+
+		count++
+		Info.Printf("Scenarios executed : %d \n =============================\n\n", count)
+	}
+}
+
+
+func (bot *uabot) continuallyUpdateTimeVisitsEvery(timeDuration time.Duration, timeVisits *int) {
+	ticker := time.NewTicker(timeDuration)
+	go func() {
+		for _ = range ticker.C {
+			var effectiveMeanTimeBetweenVisits = DEFAULTTIMEBETWEENVISITS
+			if time.Now().Weekday() == time.Saturday || time.Now().Weekday() == time.Sunday{
+				effectiveMeanTimeBetweenVisits = DEFAULTTIMEBETWEENVISITS * WEEKEND_MODIFIER
+			}
+			var randomPositiveTime int
+			for randomPositiveTime = 0; randomPositiveTime <= 0 ; randomPositiveTime = int(float64(DEFAULT_STANDARD_DEVIATION_BETWEEN_VISITS) * bot.random.NormFloat64() + 0.5) + effectiveMeanTimeBetweenVisits{}
+			*timeVisits = randomPositiveTime
+			Info.Println("Updating Time Visits to", *timeVisits)
+		}
+	}()
+}
+
+func (bot*uabot) continuallyRefreshScenariosEvery(timeDuration time.Duration, conf *Config) {
+	ticker := time.NewTicker(timeDuration)
+	go func() {
+		for _ = range ticker.C {
+			conf2 := refreshScenarios(bot.scenarioURL, !bot.local)
```

Comment:
`!bot.local` is pretty confusing here, I would keep the same boolean logic isLocal instead of isUrl.